Gems/Atom/Feature/Common/Assets/Shaders/ScreenSpace/DeferredFog.azsl

```diff
@@ -194,17 +198,32 @@ PSOutput MainPS(VSOutput IN)
         linearDepth *= lerp(1.0, fogDensity, fogDistanceDeviationRatio);
 
         // The following line is the linear depth traversing the layer
-        linearDepth = fogTraversalRelLength * linearDepth - max( 0, PassSrg::m_fogStartDistance - distToFogLayer * linearDepth);
+        linearDepth = max(0, fogTraversalRelLength * linearDepth - max( 0, PassSrg::m_fogStartDistance - distToFogLayer * linearDepth));
     }
     else
 #endif // ENABLE_FOG_LAYER
     {
         linearDepth = max( 0, linearDepth - PassSrg::m_fogStartDistance );
     }
  
-    float layerFogAmountWithStartDist = saturate(linearDepth / (PassSrg::m_fogEndDistance - PassSrg::m_fogStartDistance));
+    float density = 0.0f;
+    switch(o_fogMode)
+    {
+        case FogMode::LinearMode:
+            density = saturate(linearDepth / (PassSrg::m_fogEndDistance - PassSrg::m_fogStartDistance));
+            break;
+        case FogMode::ExponentialMode:
+            density = 1.0 - exp(-linearDepth * PassSrg::m_fogDensity); 
+            break;
+        case FogMode::ExponentialSquaredMode:
+            density = 1.0 - exp(-pow(linearDepth * PassSrg::m_fogDensity, 2));
+            break;
+        default:
+            break;
+    }
```

Comment:
Lets ask Carbonated which one they want and only add that. We are adding dynamic branches which will cause perf loss unless you build shader variants and set the shader options properly.

Reply:
Then. Lets build out the variants for o_fogMode as part of this PR and hook it up to a cvar which has a default value for mobile. That way for mobile it will always pick one option by default and not pay the cost for the others.

Reply:
ok cool yeah lets get o_fogMode added to the variant list. Also I think given the small shader and the small number of variants let's add them to the core engineer. I dont think the impact is that significant. We do this for some other core shaders too.